File: apps/agent/src/agents/workspaceAgent/projections/gmailMessage.ts
import type { MessageProjection } from '@lifecoach/shared-types';

/**
 * Project a raw `gmail.users.messages.get` (format=full) response into the
 * shape the LLM can actually read:
 *
 *   - `body.data` is base64url-encoded by the Gmail API (RFC 4648 §5).
 *     We walk the payload tree, pick the first text/plain part, fall back
 *     to text/html with a tag-strip if no plain part exists, and decode.
 *   - `payload.headers` is bloated with DKIM / ARC / Received chains we
 *     don't need. We allow-list the few that matter for triage.
 *   - The body is capped at 4 KB so a single chunky email can't blow the
 *     sub-agent's context budget.
 *
 * Pure function — the sub-agent's `get_message` tool calls into this after
 * the gws CLI returns. Same helper is reusable from any future caller.
 */

const ALLOWED_HEADERS = new Set([
  'from',
  'to',
  'cc',
  'bcc',
  'subject',
  'date',
  'reply-to',
  'list-unsubscribe',
  'message-id',
  'in-reply-to',
  'references',
]);

const BODY_BYTE_CAP = 4096;
const TRUNCATION_MARKER = '\n…[truncated]';

interface RawHeader {
  name: string;
  value: string;
}

interface RawPart {
  mimeType?: string;
  filename?: string;
  headers?: RawHeader[];
  body?: { data?: string; size?: number };
  parts?: RawPart[];
}

interface RawMessage {
  id?: string;
  threadId?: string;
  snippet?: string;
  payload?: RawPart;
}

export function projectGmailMessage(raw: RawMessage): MessageProjection {
  const payload = raw.payload ?? {};
  const headerMap = collectHeaders(payload.headers ?? []);

  const { body, truncated } = pickBody(payload);

  const projection: MessageProjection = {
    id: raw.id ?? '',
    threadId: raw.threadId ?? raw.id ?? '',
    from: headerMap.get('from') ?? '',
    subject: headerMap.get('subject') ?? '',
    date: headerMap.get('date') ?? '',
    snippet: raw.snippet ?? '',
    body,
    truncated,
  };

  const filteredHeaders = filterHeaders(headerMap);
  if (Object.keys(filteredHeaders).length > 0) {
    projection.headers = filteredHeaders;
  }

  return projection;
}

function collectHeaders(headers: RawHeader[]): Map<string, string> {
  const map = new Map<string, string>();
  for (const h of headers) {
    if (!h.name) continue;
    const key = h.name.toLowerCase();
    // First wins — RFC 5322 says repeated headers are unusual; if Gmail
    // gives us multiples, the first is typically the canonical one.
    if (!map.has(key)) {
      map.set(key, h.value ?? '');
    }
  }
  return map;
}

function filterHeaders(headerMap: Map<string, string>): Record<string, string> {
  const out: Record<string, string> = {};
  for (const [key, value] of headerMap) {
    if (!ALLOWED_HEADERS.has(key)) continue;
    // Skip the headers we've already promoted to top-level fields.
    if (key === 'from' || key === 'subject' || key === 'date') continue;
    out[canonicaliseHeaderName(key)] = value;
  }
  return out;
}

function canonicaliseHeaderName(lower: string): string {
  return lower
    .split('-')
    .map((seg) => {
      const first = seg.charAt(0);
      return first ? first.toUpperCase() + seg.slice(1) : seg;
    })
    .join('-');
}

function pickBody(payload: RawPart): { body: string; truncated: boolean } {
  const plain = findPart(payload, 'text/plain');
  if (plain) {
    return decodeAndCap(plain.body?.data ?? '');
  }
  const html = findPart(payload, 'text/html');
  if (html) {
    const decoded = decodeAndCap(html.body?.data ?? '');
    return { body: stripHtml(decoded.body), truncated: decoded.truncated };
  }
  // Top-level body (single-part messages put the body here directly).
  if (payload.body?.data) {
    return decodeAndCap(payload.body.data);
  }
  return { body: '', truncated: false };
}

function findPart(part: RawPart, mimeType: string): RawPart | null {
  if (part.mimeType === mimeType && part.body?.data) return part;
  if (!part.parts) return null;
  for (const child of part.parts) {
    const hit = findPart(child, mimeType);
    if (hit) return hit;
  }
  return null;
}

function decodeAndCap(b64url: string): { body: string; truncated: boolean } {
  if (!b64url) return { body: '', truncated: false };
  let decoded: string;
  try {
    decoded = Buffer.from(b64url, 'base64url').toString('utf8');
  } catch {
    return { body: '', truncated: false };
  }
  const buf = Buffer.from(decoded, 'utf8');
  if (buf.byteLength <= BODY_BYTE_CAP) {
    return { body: decoded, truncated: false };
  }
  return {
    body: buf.subarray(0, BODY_BYTE_CAP).toString('utf8') + TRUNCATION_MARKER,
    truncated: true,
  };
}

function stripHtml(html: string): string {
  return html
    .replace(/<style[\s\S]*?<\/style>/gi, '')
    .replace(/<script[\s\S]*?<\/script>/gi, '')
    .replace(/<[^>]+>/g, ' ')
    .replace(/&nbsp;/g, ' ')
    .replace(/&amp;/g, '&')
    .replace(/&lt;/g, '<')
    .replace(/&gt;/g, '>')
    .replace(/&quot;/g, '"')
    .replace(/&#39;/g, "'")
    .replace(/\s+/g, ' ')
    .trim();
}

File: apps/agent_py/src/lifecoach_agent/workspace_agent/projections/gmail_message.py
"""Project a raw `gmail.users.messages.get` (format=full) response into
the shape the LLM can actually read:

- `body.data` is base64url-encoded by the Gmail API (RFC 4648 §5). We
  walk the payload tree, pick the first text/plain part, fall back to
  text/html with a tag-strip if no plain part exists, and decode.
- `payload.headers` is bloated with DKIM / ARC / Received chains we
  don't need. We allow-list the few that matter for triage.
- The body is capped at 4 KB so a single chunky email can't blow the
  sub-agent's context budget.
"""

from __future__ import annotations

import base64
import re
from typing import Any

from lifecoach_agent.contracts.models import MessageProjection

_ALLOWED_HEADERS: frozenset[str] = frozenset(
    {
        "from",
        "to",
        "cc",
        "bcc",
        "subject",
        "date",
        "reply-to",
        "list-unsubscribe",
        "message-id",
        "in-reply-to",
        "references",
    }
)

BODY_BYTE_CAP = 4096
_TRUNCATION_MARKER = "\n…[truncated]"


def project_gmail_message(raw: dict[str, Any]) -> MessageProjection:
    payload = raw.get("payload") or {}
    header_map = _collect_headers(payload.get("headers") or [])
    body, truncated = _pick_body(payload)

    filtered_headers = _filter_headers(header_map)
    return MessageProjection.model_validate(
        {
            "id": raw.get("id") or "",
            "threadId": raw.get("threadId") or raw.get("id") or "",
            "from": header_map.get("from", ""),
            "subject": header_map.get("subject", ""),
            "date": header_map.get("date", ""),
            "snippet": raw.get("snippet") or "",
            "body": body,
            "truncated": truncated,
            **({"headers": filtered_headers} if filtered_headers else {}),
        }
    )


def _collect_headers(headers: list[dict[str, Any]]) -> dict[str, str]:
    out: dict[str, str] = {}
    for h in headers:
        name = h.get("name")
        if not name:
            continue
        key = name.lower()
        # First wins — RFC 5322 says repeated headers are unusual; the
        # first is typically the canonical one when Gmail returns multiples.
        if key not in out:
            out[key] = h.get("value") or ""
    return out


def _filter_headers(header_map: dict[str, str]) -> dict[str, str]:
    out: dict[str, str] = {}
    for key, value in header_map.items():
        if key not in _ALLOWED_HEADERS:
            continue
        # Skip the headers we've already promoted to top-level fields.
        if key in {"from", "subject", "date"}:
            continue
        out[_canonicalise_header_name(key)] = value
    return out


def _canonicalise_header_name(lower: str) -> str:
    return "-".join(seg[:1].upper() + seg[1:] if seg else seg for seg in lower.split("-"))


def _pick_body(payload: dict[str, Any]) -> tuple[str, bool]:
    plain = _find_part(payload, "text/plain")
    if plain is not None:
        return _decode_and_cap((plain.get("body") or {}).get("data") or "")
    html = _find_part(payload, "text/html")
    if html is not None:
        decoded, was_truncated = _decode_and_cap((html.get("body") or {}).get("data") or "")
        return _strip_html(decoded), was_truncated
    top = payload.get("body") or {}
    if top.get("data"):
        return _decode_and_cap(top["data"])
    return "", False


def _find_part(part: dict[str, Any], mime: str) -> dict[str, Any] | None:
    if part.get("mimeType") == mime and (part.get("body") or {}).get("data"):
        return part
    for child in part.get("parts") or []:
        hit = _find_part(child, mime)
        if hit is not None:
            return hit
    return None


def _decode_and_cap(b64url: str) -> tuple[str, bool]:
    if not b64url:
        return "", False
    try:
        # urlsafe_b64decode tolerates missing padding once we re-pad to
        # a multiple of 4 — Gmail's base64url drops trailing `=`.
        pad = (-len(b64url)) % 4
        raw = base64.urlsafe_b64decode(b64url + ("=" * pad))
    except (ValueError, base64.binascii.Error):  # type: ignore[attr-defined]
        return "", False
    decoded = raw.decode("utf-8", errors="replace")
    encoded = decoded.encode("utf-8")
    if len(encoded) <= BODY_BYTE_CAP:
        return decoded, False
    capped = encoded[:BODY_BYTE_CAP].decode("utf-8", errors="ignore")
    return capped + _TRUNCATION_MARKER, True


_STYLE_RE = re.compile(r"<style[\s\S]*?</style>", re.IGNORECASE)
_SCRIPT_RE = re.compile(r"<script[\s\S]*?</script>", re.IGNORECASE)
_TAG_RE = re.compile(r"<[^>]+>")
_WHITESPACE_RE = re.compile(r"\s+")
_ENTITY_MAP = {
    "&nbsp;": " ",
    "&amp;": "&",
    "&lt;": "<",
    "&gt;": ">",
    "&quot;": '"',
    "&#39;": "'",
}


def _strip_html(html: str) -> str:
    out = _STYLE_RE.sub("", html)
    out = _SCRIPT_RE.sub("", out)
    out = _TAG_RE.sub(" ", out)
    for entity, replacement in _ENTITY_MAP.items():
        out = out.replace(entity, replacement)
    return _WHITESPACE_RE.sub(" ", out).strip()
